xarray_beam/_src/pangeo_forge.py

```diff
@@ -89,62 +107,38 @@ def __init__(
     self.sub_chunks = sub_chunks or -1
     self.xarray_open_kwargs = xarray_open_kwargs or {}
 
-  def _prechunk(self) -> Iterator[Tuple[core.Key, Tuple[int, ...]]]:
-    """Converts `FilePattern` items into keyed indexes."""
-    input_chunks = {k: v or 1 for k, v in self.pattern.nitems_per_input.items()}
-    dim_sizes = {
-      k: v or self.pattern.dims[k]
-      for k, v, in self.pattern.concat_sequence_lens.items()
-    }
-    chunks = core.normalize_expanded_chunks(input_chunks, dim_sizes)
-    for key, (index, _) in zip(core.iter_chunk_keys(chunks),
-                               self.pattern.items()):
-      yield key, index
-
-  def _open_chunks(
-      self,
-      key: core.Key,
-      index: Tuple[int, ...]
-  ) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
+    if pattern.merge_dims:
+      raise ValueError("patterns with `MergeDim`s are not supported.")
+
+  def _open_chunks(self, _) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
     """Open datasets into chunks with XArray."""
-    path = self.pattern[index]
-    with FileSystems().open(path) as file:
-      dataset = xarray.open_dataset(
-        file, chunks=self.sub_chunks, **self.xarray_open_kwargs
-      )
-      dataset = _expand_dimensions_by_key(dataset, key, index, self.pattern)
+    for index, path in self.pattern.items():
+      with FileSystems().open(path) as file:
+        key = _pattern_index_to_key(index)
 
-      base_key = core.Key(_zero_dimensions(dataset)).with_offsets(**key.offsets)
+        dataset = xarray.open_dataset(
+          file, chunks=self.sub_chunks, **self.xarray_open_kwargs
+        )
+        dataset = _expand_dimensions_by_key(dataset, key, index, self.pattern)
 
-      num_threads = len(dataset.data_vars)
+        base_key = core.Key(_zero_dimensions(dataset)).with_offsets(
+          **key.offsets
+        )
 
-      if self.sub_chunks == -1:
-        yield base_key, dataset.compute(num_workers=num_threads)
-        return
+        num_threads = len(dataset.data_vars)
 
-      dim_sizes = {dim: dataset.dims[dim] for dim in self.sub_chunks.keys()}
-      norm_sub_chunks = core.normalize_expanded_chunks(
-        self.sub_chunks, cast(Mapping[str, int], dim_sizes)
-      )
-      offset_index = core.compute_offset_index(
-        core._chunks_to_offsets(norm_sub_chunks)
-      )
-      for sub_key in core.iter_chunk_keys(norm_sub_chunks):
-        sizes = {
-          dim: norm_sub_chunks[dim][offset_index[dim][offset]]
-          for dim, offset in sub_key.offsets.items()
-        }
-        slices = core.offsets_to_slices(
-          sub_key.offsets, sizes=sizes, base=base_key.offsets
-        )
-        chunk = dataset.isel(slices)
+        if self.sub_chunks == -1:
+          yield base_key, dataset.compute(num_workers=num_threads)
+          return
```

Comment:
Does this special case actually make a meaningful difference? If not, I would suggest removing it.

Reply:
Yes, this fits a distinct use case. I've added a comment about the if to make it clearer to the reader.